webui/react/src/pages/TaskList.tsx

```diff
@@ -108,6 +113,66 @@ const TaskList: React.FC = () => {
     triggerTensorboardsRequest,
   ]);
 
+  const handleSourceExpand = useCallback((event: React.MouseEvent, id: string) => {
+    event.stopPropagation();
+    setSourceExpanded(prev => {
+      const newSourceExpanded = { ...prev };
+      if (!newSourceExpanded[id]) newSourceExpanded[id] = true;
+      else delete newSourceExpanded[id];
+      return newSourceExpanded;
+    });
+  }, []);
+
+  const columns = useMemo(() => {
+    const sourceRenderer: TaskRenderer = (_, record) => {
+      const info = {
+        isPlural: false,
+        label: '',
+        path: '',
+        source: [] as number[],
+      };
+      if (record.misc?.experimentIds) {
+        info.label = 'Experiment';
+        info.path = '/ui/experiments';
+        info.source = record.misc.experimentIds || [];
+      } else if (record.misc?.trialIds) {
+        info.label = 'Trial';
+        info.path = '/ui/trials';
+        info.source = record.misc.trialIds || [];
+      }
+      info.isPlural = info.source.length > 1;
+      info.source.sort(numericSorter);
+
+      const isExpanded = sourceExpanded[record.id];
+      const sourceCount = info.source.length;
+      const showToggle = info.source.length > MAX_SOURCES;
+      const toggleLabel = `${isExpanded ? 'Hide' : 'Show'} ${sourceCount}${isExpanded ? '' : '+'}`;
```

Comment:
Blocking: The behavior of this label isn't quite correct.

When I have 8 experiments, it shows the first 5 source and has a button saying `Show 8+`. This isn't intuitive to me, since clicking on it will show 3 more. I think the `+` should disappear and we might want to think about rephrasing to something like `Show all 8` or `Show 3 more`.

After I have expanded, it says `Hide 8`. However, clicking on that only hides 3 sources.

Reply:
When there are 8, and set to only show 5 by default, I've updated it to show `Show 3+` and `Hide 3`.

Reply:
Thought about it more and I think `Show 3+` and `Collapse` makes more sense. `Hide X` seems arbitrary and the number to hide is not useful and just causes confusion.